### PR TITLE
[BLOG] Neural Sparse is now available in HuggingFace SentenceTransformers

### LINKED ISSUE #3798: [BLOG] Neural Sparse is now available in HuggingFace SentenceTransformers
### Describe the blog post

We have collaborated with SentenceTransformers on building code examples for using neural sparse in OpenSearch; as well as building an out-of-box user experience for inf-free sparse models.

This blog post describes how to use SentenceTransformers sparse encoders with OpenSearch, and OpenSearch's tech innovations in it.

### Expected Title

Neural Sparse is now available in HuggingFace SentenceTransformers

### Authors Name

Zhichao Geng

### Authors Email

zhichaog@amazon.com

### Target Draft Date

06/09/2025

### Blog Post Category

technical

### Target Publication Date

_No response_

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by zhichao-aws**
### Description
add blog **Neural Sparse is now available in HuggingFace SentenceTransformers**.
 
### Issues Resolved
#3798

### Check List
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

### CODE REVIEW

**Comment by zhichao-aws** [comment]
Mark the status as draft, we still need to modify the links in this blog after SentenceTransformers actually release it.

**Review by arthurbr11 — COMMENTED** [review]
Thanks a lot for the blog post ! I added some small reviews but nitpick mostly. Will just need to be updating the link but you mentioned it already so should be all good ! 
cc @tomaarsen

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+Compared to dense embedding models, sparse models have a more complex structure, requiring post-processing layers for vector sparsification and complex tensor-to-map conversions for OpenSearch compatibility. These factors collectively increase the technical barrier to using and deploying neural sparse models in Python.
+
+Sentence Transformers offers a streamlined user experience. All encoding details of OpenSearch neural sparse models are abstracted away. Configuration aspects like bi-encoder or doc-only (inference-free) mode and activation functions are handled internally. Model initialization and text encoding require just 1-2 lines of code.
+
+Now let's explore neural sparse in Sentence Transformers with hands-on coding examples!
+
+### Install python dependencies
+```
+pip install sentence-transformers==4.2.0
```
[the remark this answers is not in the document]
Comment (zhichao-aws) [reply _posts/2025-06-11-neural-sparse-sentence-transformers.md:38]: ack

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+---
+
+OpenSearch's neural sparse feature transforms text into sparse token-weight pairs using transformer models, combining semantic search capabilities with efficient inverted indexing. This approach delivers high retrieval accuracy with low latency and minimal resource usage, offering a scalable alternative to dense vector search. The inference-free mode eliminates query encoding overhead, enabling extremely fast search. OpenSearch provides state-of-the-art pre-trained models that consistently lead the [BEIR](https://github.com/beir-cellar/beir) benchmark for sparse retrievers, and has yield over **10M+ downloads** on Hugging Face!
+
+The [Sentence Transformers](https://huggingface.co/sentence-transformers)(a.k.a. SBERT) library, developed by [UKPLab](https://www.tu-darmstadt.de/) and maintained by [Hugging Face](https://huggingface.co/), is a Python framework designed to generate semantically meaningful embeddings for sentences, paragraphs, and images. The library is now widely adopted in the natural language processing (NLP) tasks such as semantic search, clustering, paraphrase mining, and retrieval-augmented generation.
```
[the remark this answers is not in the document]
Comment (zhichao-aws) [reply _posts/2025-06-11-neural-sparse-sentence-transformers.md]: ack

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+### Encoding Examples
+
+The model supports encoding for both documents and queries.
+
+#### Query Encoding
+```python
+# We first use encode method to convert text to torch.sparse_coo tensor (it's inference-free here)
+# Then we use decode method to convert torch.sparse_coo tensor to (token, weight) pairs
+
+query_tensor = sparse_model.encode([{"query": "What's the weather in ny now?"}])
```
[the remark this answers is not in the document]
Comment (zhichao-aws) [reply _posts/2025-06-11-neural-sparse-sentence-transformers.md:62]: Yes let's update it after merge

**Comment by zhichao-aws** [comment]
Hi @pajuric , the blog content is done, we're pending SentenceTransformers release now (may have some changes on the sample python code). 

Can we start the review now?

**Comment by kolchfa-aws** [comment]
@zhichao-aws Doc review complete. Please review the edits and verify that everything looks good, and we'll move on to editorial review.

**Comment by zhichao-aws** [comment]
Thanks @kolchfa-aws . The changes look good to me.

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+First, install the required Python library:
+
+```bash
+pip install sentence-transformers==5.0.0
+```
+
+### Step 2: Initialize a neural sparse model
+
+Import the SparseEncoder class and load a pretrained model. You can find more models in OpenSearch's [Hugging Face repo](https://huggingface.co/opensearch-project):
```
Comment (natebower) [inline _posts/2025-06-11-neural-sparse-sentence-transformers.md:46]:
```suggestion
Import the SparseEncoder class and load a pretrained model. You can find more models in OpenSearch's [Hugging Face repository](https://huggingface.co/opensearch-project):
```

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+You can also explore example scripts for working with the following embedding types:
+
+- [Sparse embeddings](https://github.com/UKPLab/sentence-transformers/blob/master/examples/sparse_encoder/applications/semantic_search/semantic_search_opensearch.py)
+- [Dense embeddings](https://github.com/UKPLab/sentence-transformers/blob/master/examples/sentence_transformer/applications/semantic-search/semantic_search_nq_opensearch.py)
+
+
+## OpenSearch's contributions to inference-free models in Sentence Transformers
+
+The OpenSearch team is leading innovation in learned sparse retrieval. Our research in inference-free sparse encoders [^1][^2] is increasingly adopted as standard practice across the industry. The following techniques have now been integrated into the Sentence Transformers library, making them more accessible to the broader community.
```
Comment (natebower) [inline _posts/2025-06-11-neural-sparse-sentence-transformers.md:138]:
```suggestion
OpenSearch is driving innovation in learned sparse retrieval. Our research into inference-free sparse encoders [^1][^2] is increasingly adopted as standard practice across the industry. The following techniques have now been integrated into the Sentence Transformers library, making them more accessible to the broader community.
```

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+### IDF-aware Penalty: Enhancing Relevance and Efficiency
+
+Traditional sparse retrievers apply uniform penalties to all tokens during training, ignoring the varying importance of different terms. To address this limitation, OpenSearch researchers introduced IDF-aware penalty[^1], which modulates the penalty based on token significance. This approach applies stronger penalties to common tokens while preserving rare, information-rich tokens that carry more distinctive meaning.
+
```
Comment (natebower) [inline _posts/2025-06-11-neural-sparse-sentence-transformers.md:143]: Line 142: Either "an" or "the" should precede "IDF-aware penalty". The 2nd sentence reads as repetitive. An/the IDF-aware penalty modulates the penalty? Can we clarify?

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+With this enhancement, inference-free retrievers significantly outperform baselines on zero-shot retrieval. The IDF-aware penalty also improves search efficiency by reducing average FLOPs on inverted indexes. Importantly, this enhancement serves as a prerequisite for implementing the powerful [two-phase search](https://opensearch.org/blog/Introducing-a-neural-sparse-two-phase-algorithm) methodology. All inference-free models[^3] released by our team incorporate this enhancement.
```
Comment (natebower) [inline _posts/2025-06-11-neural-sparse-sentence-transformers.md:144]: Last sentence: "our team" => "Hugging Face"?

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+sparse_model = SparseEncoder(
+    modules=[router],
+    similarity_fn_name="dot",
+)
+```
+
+### ℓ0-inspired sparsification: Advancing sparse retrieval
+
+Achieving optimal balance between sparsity and relevance has been a persistent challenge for inference-free retrievers. The OpenSearch team addressed this by developing two novel ℓ0-inspired sparsification approaches[^2] (accepted to SIGIR 2025): ℓ0 mask loss, which excludes already-sparse documents from further regularization, and ℓ0 Approximation Activation, which uses log transformations to alter the penalty on tokens with lower activations.
```
Comment (natebower) [inline _posts/2025-06-11-neural-sparse-sentence-transformers.md:179]:
```suggestion
Achieving optimal balance between sparsity and relevance has been a persistent challenge for inference-free retrievers. OpenSearch addressed this by developing two novel ℓ0-inspired sparsification approaches[^2] (accepted to SIGIR 2025): ℓ0 mask loss, which excludes already-sparse documents from further regularization, and ℓ0 Approximation Activation, which uses log transformations to alter the penalty on tokens with lower activations.
```

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+Models using these ℓ0-inspired techniques demonstrate superior performance while significantly improving efficiency and sparsity. The ℓ0 mask loss also streamlines the training process by eliminating the need for multiple attempts at tuning the FLOPS lambda hyperparameter.
+Our [doc-v3-distill](https://huggingface.co/opensearch-project/opensearch-neural-sparse-encoding-doc-v3-distill) and [doc-v3-gte](https://huggingface.co/opensearch-project/opensearch-neural-sparse-encoding-doc-v3-gte) models incorporate these enhancements, delivering state-of-the-art performance.
```
Comment (natebower) [inline _posts/2025-06-11-neural-sparse-sentence-transformers.md:182]:
```suggestion
Hugging Face's [doc-v3-distill](https://huggingface.co/opensearch-project/opensearch-neural-sparse-encoding-doc-v3-distill) and [doc-v3-gte](https://huggingface.co/opensearch-project/opensearch-neural-sparse-encoding-doc-v3-gte) models incorporate these enhancements, delivering state-of-the-art performance.
```

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+```
+
+## Further reading
+
+For more information about neural sparse search, see these previous blog posts:
+
+- [Improving document retrieval with sparse semantic encoders](https://opensearch.org/blog/improving-document-retrieval-with-sparse-semantic-encoders)
+- [A deep dive into faster semantic sparse retrieval in OpenSearch 2.12](https://opensearch.org/blog/A-deep-dive-into-faster-semantic-sparse-retrieval-in-OS-2.12/)
+- [Introducing the neural sparse two-phase algorithm](https://opensearch.org/blog/Introducing-a-neural-sparse-two-phase-algorithm)
+- [Advancing Search Quality and Inference Speed with v2 Series Neural Sparse Models](https://opensearch.org/blog/neural-sparse-v2-models)
```
Comment (natebower) [inline _posts/2025-06-11-neural-sparse-sentence-transformers.md:205]:
```suggestion
- [Advancing search quality and inference speed with v2 series neural sparse models](https://opensearch.org/blog/neural-sparse-v2-models)
```

**Review by natebower — COMMENTED** [review]
@zhichao-aws Editorial review complete. Please see my comments and changes and let me know if you have any questions. Thanks!

**Comment by natebower** [comment]
@pajuric This will be ready for you once @zhichao-aws addresses my comments/changes. Thanks!

**Comment by pajuric** [comment]
@zhichao-aws - Blog is scheduled to publish on 7/3 pending implementation and finalization of @natebower's edits.

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+You can also explore example scripts for working with the following embedding types:
+
+- [Sparse embeddings](https://github.com/UKPLab/sentence-transformers/blob/master/examples/sparse_encoder/applications/semantic_search/semantic_search_opensearch.py)
+- [Dense embeddings](https://github.com/UKPLab/sentence-transformers/blob/master/examples/sentence_transformer/applications/semantic-search/semantic_search_nq_opensearch.py)
+
+
+## OpenSearch's contributions to inference-free models in Sentence Transformers
+
+The OpenSearch team is leading innovation in learned sparse retrieval. Our research in inference-free sparse encoders [^1][^2] is increasingly adopted as standard practice across the industry. The following techniques have now been integrated into the Sentence Transformers library, making them more accessible to the broader community.
+
```
[the remark this answers is not in the document]
Comment (zhichao-aws) [reply _posts/2025-06-11-neural-sparse-sentence-transformers.md:139]: Our intention is to refer to OpenSearch. Because those 2 models are trained by OpenSearch and released at Hugging Face. Let me change it to OpenSearch.

File: _posts/2025-06-11-neural-sparse-sentence-transformers.md (hunk 7 above)
Comment (zhichao-aws) [reply _posts/2025-06-11-neural-sparse-sentence-transformers.md:144]: it's OpenSearch. I'll change it

**Review by kolchfa-aws — APPROVED** [review]
